[Vega] Implement context filter modification

nyurik opened the pull request:
Implements https://github.com/elastic/kibana/issues/17210

API:
* `kibanaAddFilter` - adds ES query filter to the filter bar
```js
kibanaAddFilter(
  { "match": { "extension.keyword": { "query": "jpg", "type": "phrase"  }}}   // ES Query language
  { } // Optional filter metadata, e.g. enabled/disabled, etc
)
```
* `kibanaRemoveFilter(query, [meta])` - same params as adding
* `kibanaRemoveAllFilters()`
* `kibanaSetTimeFilter(start, end)` sets time range for the dashboard - works with numbers, strings, and dates.

Extensive example:  https://gist.github.com/nyurik/3c579fb8e81c8a981d4b267ad1767227
* select portion of the graph, move that portion around, resize on each side by the edge handles, and either double click or click apply to set the time.
* click any of the extensions to add a filter
* click on the extension again (when there is only one) to remove that filter
* double click the graph to reset the time to last 15 days

![screenshot from 2018-04-09 08-37-13](https://user-images.githubusercontent.com/1641515/38481022-778c69f0-3bd1-11e8-8d80-0fe1081148cf.png)



Simple testing code (click or shift click to set/unset)

```hjson
{
  $schema: https://vega.github.io/schema/vega/v3.json
  title: click anywhere to set, shift click to remove
  signals: [
    {
      name: myFilter
      value: {
        match: {
          extension.keyword: {query: "jpg", type: "phrase"}
        }
      }
    }
    {
      name: btn
      on: [
        {events: "view:click[!event.shiftKey]", update: "kibanaAddFilter(myFilter)"}
        {events: "view:click[event.shiftKey]", update: "kibanaRemoveFilter(myFilter)"}
      ]
    }
  ]
}
```

Release Summary: Vega now allows modifying filters and the time range in Kibana, by adding utility functions, that can be called in your signal handlers. Check the [PR description](https://github.com/elastic/kibana/pull/17586) for a description of the functions.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1482/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1510/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1531/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1535/)

alexfrancoeur commented:
@nyurik I tested this out a bit. Didn't have time to add a filter to an already existing chart, but it does seem to work great. Filters are applied to the dashboard and update all other visualizations appropriately. My only feedback is around the "apply filter" button after selecting rom a time range. Is there any reason why we don't automatically apply the filter if you've selected a timeframe? That would match all other charts. Can this be done or was it the way the example you provided was configured?

<img width="1257" alt="screen shot 2018-04-09 at 8 07 08 pm" src="https://user-images.githubusercontent.com/25181823/38529240-a14d9c9a-3c31-11e8-826c-5b627d5fdc3f.png">

nyurik commented:
@alexfrancoeur that's just how I designed that specific graph - I wanted to be able to readjust the selected range -- dragging the sides or the whole thing, before confirming (button click or range doubleclick) that's the range I want. But from Vega's perspective, you can do these commands at any point, even on load - e.g. if the data contains something bad, zoom in right away.

nyurik commented:
@bargs, could you take a look at the filter approach I'm using. (@alexfrancoeur suggested i poke you on this)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1659/)

nyurik commented:
jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1662/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1711/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1713/)

timroes commented:
I have two topics to discuss around this PR. @alexfrancoeur I would like your feedback on both of them.

#### Error handling

You can very easily create invalid filters that way, that will just not appear in the filter bar, but just throw an error in the console, e.g. if you use a field name that doesn't exist on the (specified or default) index pattern.

In this cases there just won't happen anything except the exception on the console. We cannot properly catch this error (like with a `try-catch`), but to validate them we would need to really validate input parameters beforehand (like checking that the field name exists, checking if the operator is a valid one). In that case we could show an error toast.

The question is, how do we want to handle that kind of errors. On the one hand Vega is for expert users, and we kind of could request from them to be able to not create invalid parameters or are able to check the browser dev tools for errors. On the other hand, this has two drawbacks: 

* A user using the visualization might just be able to start an interaction without getting any feedback that anything failed, which can be confusing (but validating all input parameters is also a bit more code in this place).
* We will require users to disable all vega visualizations before they try to report issues here or in the forums, because they are seeing errors in the browser console. If we give the user that direct access to APIs where we know they can easily create a Vega spec, that throws errors on the console, we cannot accept any console error issues anymore as long as Vega is running, since we never can rule out, the person just had some broken Spec on the dashboard.

So we are basically down to these two options: not validating any inputs, with the two drawbacks above, or spending way more effort validating, that the parameters in the spec (like field names, operators, index pattern, etc.) are valid and in this case since we do all the validation in that place, we could also show toast notifications for failures.

#### Setting time ranges

I think we can make setting the time range more safe AND user friendly :-) I would get rid of explicitly specifying the mode altogether. There are basically only two modes, that make sense: `absolute` and `relative` (since every `quick` range is just a `relative` range).

I would autodetect which one we should be using, and here are the ideas for how the function should behave (in all cases I would check, that `from` and `to` are in the correct order, i.e. switching them if `to` is before `from`):

* If both are valid dates, I would use `absolute` mode setting these dates.
* If both are strings, pass them to [`datemath.parse`](https://github.com/elastic/datemath-js/blob/master/src/index.js) (which converts `now-7d`, etc. to moment instances) and check if the returned `moment` instance is a valid instance. If so use `relative` mode with the original strings.
* If one is a valid date and one is a string, check if `datemath.parse` can parse the string. If so, use `absolute` mode and convert the string at time of clicking to the absolute date.
* In all other cases show a toast notification with an error, that the specified time range was invalid.

I would suggest an implementation like the following:

```js
import dateMath from '@kbn/datemath';

function setTimeRange(from, to, mode) {
	this._timefilter.from = from;
	this._timefilter.to = to;
	this._timefilter.mode = mode;
	this._timefilter.update();
}

view.kibanaSetTimeFilterHandler = (start, end) => {
	const fromDate = moment(start);
	const toDate = moment(end);

	if (fromDate.isValid() && toDate.isValid()) {
		// Both are valid dates, use them absolute
		setTimeRange(fromDate, toDate, 'absolute');
	} else if (from.isValid() || to.isValid()) {
		// Only one is valid, we need to check if the other is a valid datemath string.
		// Since datemath.parse will also convert absolute date strings, we can just convert
		// both here.
		const from = dateMath.parse(start);
		const to = dateMath.parse(end);
		if (!from.isValid() || !to.isValid()) {
			// One was an absolute string the other NOT a valid datemath string.
			// TODO: show error toast
			return;
		}
		setTimeRange(from, to, 'absolute');
	} else if (typeof start === 'string' && typeof end === 'string') {
		const from = dateMath.parse(start);
		const to = dateMath.parse(end);
		if (!from.isValid() || !to.isValid()) {
			// one was not a valid datemath string
			// TODO: show error toast
			return;
		}
		setTimeRange(from, to, 'relative');
	} else {
		// TODO: Show error toast
	}
};
```

timroes commented on src/core_plugins/vega/public/vega_view/vega_base_view.js:
```diff
@@ -32,6 +59,14 @@ export class VegaBaseView {
 
     try {
+      // Adapted from src/ui/public/filter_editor/filter_editor.js
+      if (!areIndexPatternsProvided(this._indexPatterns)) {
```
This does properly not what you want. Where should `this._indexPatterns` come from in this place? This will always be the index patterns Angular service in your case? We need a way to specify the index pattern via the function because at the moment it will always create the filter on the default index pattern.

We should either accept an index pattern id to the `kibanaAddFilterHandler` method or alternative I think a index pattern itself, and we try to find that index pattern in the list of all index patterns.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2042/)

ppisljar commented on src/core_plugins/vega/public/vega_visualization.js:
```diff
@@ -1,10 +1,12 @@
 import { Notifier } from 'ui/notify';
 import { VegaView } from './vega_view/vega_view';
 import { VegaMapView } from './vega_view/vega_map_view';
+import { FilterBarQueryFilterProvider } from 'ui/filter_bar/query_filter';
```
you can get this on the `vis` object (`vis.API.queryFilter`)

ppisljar commented on src/core_plugins/vega/public/vega_visualization.js:
```diff
@@ -1,10 +1,12 @@
+import { FilterBarQueryFilterProvider } from 'ui/filter_bar/query_filter';
 
-export function VegaVisualizationProvider(vegaConfig, serviceSettings) {
+export function VegaVisualizationProvider(Private, vegaConfig, serviceSettings, indexPatterns, timefilter) {
```
get the indexPatterns and timefilter from `vis.API.indexPatterns` and `vis.API.timeFilter`

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2321/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2356/)

nyurik commented:
@timroes @ppisljar thx for your feedback, I think I fixed all the issues you have raised. I added standard Vega error handling, and added an optional index parameter. At the moment, I use `findObjectByTitle(savedObjectsClient, 'index-pattern', index)` - seems to be working.  It could be improved with caching and moving this code to index pattern class, but I am not sure how core team wants to handle this.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2715/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2723/)

timroes commented on src/core_plugins/vega/public/vega_view/vega_base_view.js:
```diff
@@ -144,16 +180,109 @@ export class VegaBaseView {
+  async addFilterHandler(query, index) {
+    const indexId = await this._findIndex(index);
+    const filter = buildQueryFilter(query, indexId);
+    this._queryfilter.addFilters(filter);
```
`addFilters` actually returns a promise, that will fail if the filter was invalid. I think we should use this here to show an error message to the user.

```js
try {
  await this._queryfilter.addFilters(filter);
} catch {
  throw new Error('Cannot add filter, since it was invalid.'); // Or another similar error message
}
```

timroes commented:
Just as a to do list, before we can continue on this one:

* Change the time range setting behavior as mentioned in [the comment above](https://github.com/elastic/kibana/pull/17586#issuecomment-382300244)
* #18863 needs to be solved (not within this PR, and not an issue of you, we just need to figure out the best way to solve it).
* Some minor changes in error handling (I commented inline)

vcaldarelli commented:
Hi to all.
One question.
I tried the example in [](https://gist.github.com/nyurik/3c579fb8e81c8a981d4b267ad1767227) in Kibana 6.2.4 but I got the error "Unrecognized function: kibanaSetTimeFilter".

nyurik commented:
@vcaldarelli this pull request hasn't been merged yet.

vcaldarelli commented:
@nyurik thanks!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5433/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5529/)

nyurik commented:
@timroes @ppisljar I think I addressed all the raised issues. Please review. I have been getting too many requests for this feature to ignore it any longer on my part :)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5666/)

nyurik commented:
jenkins retest this

nyurik commented:
jenkins, what's wrong with you? retest this!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5691/)

ppisljar commented:
the tests are broken, i have a PR open https://github.com/elastic/kibana/pull/20287

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5732/)

timroes commented on src/core_plugins/vega/public/vega_view/vega_base_view.js:
```diff
@@ -188,6 +222,115 @@ export class VegaBaseView {
+      // Try to parse as relative dates too (absolute dates will also be accepted)
+      const startDate = dateMath.parse(start);
+      const endDate = dateMath.parse(end);
+      if (!startDate.isValid() || !endDate.isValid()) {
```
`startDate` and `endDate` could be `undefined` in case the user specified an invalid format. I think we should fix that `if` for also checking for that.

nyurik replied:
done. BTW, note that `moment({a:1})` is, for some weird reason, is a valid date :(

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5854/)

nyurik commented on src/core_plugins/vega/public/vega_view/vega_base_view.js:
```diff
@@ -188,6 +222,115 @@ export class VegaBaseView {
+        // in case functions don't match the list above
+        throw new Error(`${funcName}() is not defined for this graph`);
+      }
+      await this[handlerFunc].apply(this, args);
```
good catch!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5914/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5920/)

nicpenning commented on src/core_plugins/vega/public/vega_type.js:
```diff
@@ -60,8 +60,8 @@ VisTypesRegistryProvider.register((Private) => {
       showIndexSelection: false,
-      showQueryBar: false,
-      showFilterBar: false,
+      showQueryBar: true,
```
Will this apply to the OSS version?

acro-yoshioka commented:
> Implements #17210
> 
> API:
> 
> * `kibanaAddFilter` - adds ES query filter to the filter bar
> 
> ```js
> kibanaAddFilter(
>   { "match": { "extension.keyword": { "query": "jpg", "type": "phrase"  }}}   // ES Query language
>   { } // Optional filter metadata, e.g. enabled/disabled, etc
> )
> ```

Please tell me how to specify meta.
I would like to specify the label of the Content Filter.

nyurik commented:
@acro-yoshioka I don't think it is possible at the moment, please create a new issue for that. CC: @timroes

acro-yoshioka commented:
> @acro-yoshioka I don't think it is possible at the moment, please create a new issue for that. CC: @timroes

Thank you for your comment !
I understand that the 2nd arg about Optional filter metadata is not implemented.

francisca-lima commented:
Any solution yet? Thank you!

stephlane544 commented:
Is it possible to use this dynamically? I looking to add a filter based on the data point you click on, but am having issues determining what to insert into the kibanaAddFilter to get it to actually use information from the data point I clicked on...
